cpp/oneapi/dal/backend/primitives/sort/sort_dpc.cpp

```diff
@@ -0,0 +1,540 @@
+                              ndarray<IndexType, 1>& part_hist,
+                              IndexType elem_count,
+                              std::uint32_t bit_offset,
+                              std::int64_t local_size,
+                              std::int64_t local_hist_count) {
+    ONEDAL_ASSERT(part_hist.get_count() == ((local_hist_count + 1) << radix_bits));
+
+    sycl::range<1> global(local_size * local_hist_count);
+    sycl::range<1> local(local_size);
+    sycl::nd_range<1> nd_range(global, local);
+
+    const RadixIntType* val_ptr =
+        static_cast<const RadixIntType*>(static_cast<const void*>(val.get_data()));
+    IndexType* part_hist_ptr = part_hist.get_mutable_data();
+
+    auto event = queue.submit([&](cl::sycl::handler& cgh) {
+        cgh.parallel_for(nd_range, [=](sycl::nd_item<1> item) {
+            auto sbg = item.get_sub_group();
+            if (sbg.get_group_id()[0] > 0) {
+                return;
+            }
+            const std::uint32_t n_groups = item.get_group_range(0);
+            const std::uint32_t n_sub_groups = sbg.get_group_range()[0];
+            const std::uint32_t n_total_sub_groups = n_sub_groups * n_groups;
+            const IndexType elems_for_sbg =
+                elem_count / n_total_sub_groups + !!(elem_count % n_total_sub_groups);
+            const std::uint32_t local_size = item.get_sub_group().get_local_range()[0];
+
+            const std::uint32_t local_id = item.get_sub_group().get_local_id()[0];
+            const std::uint32_t sub_group_id = sbg.get_group_id();
+            const std::uint32_t group_id = item.get_group().get_id(0) * n_sub_groups + sub_group_id;
+
+            IndexType ind_start = group_id * elems_for_sbg;
+            IndexType ind_end = (group_id + 1) * elems_for_sbg;
+
+            if (ind_end > elem_count) {
+                ind_end = elem_count;
+            }
+
+            IndexType offset[(std::uint32_t)1 << radix_bits];
+            const std::uint32_t radix_range = (std::uint32_t)1 << radix_bits;
+            const std::uint32_t radix_range_1 = radix_range - 1;
+            for (std::uint32_t i = 0; i < radix_range; i++) {
+                offset[i] = 0;
+            }
+
+            for (IndexType i = ind_start + local_id; i < ind_end; i += local_size) {
+                RadixIntType data_bits = ((inv_bits(val_ptr[i]) >> bit_offset) & radix_range_1);
+                for (std::uint32_t j = 0; j < radix_range; j++) {
+                    IndexType value = static_cast<IndexType>(data_bits == j);
+                    IndexType partial_offset =
+                        sycl::ONEAPI::reduce(sbg, value, cl::sycl::ONEAPI::plus<IndexType>());
+                    offset[j] += partial_offset;
+                }
+            }
+
+            if (local_id == 0) {
+                for (std::uint32_t j = 0; j < radix_range; j++) {
+                    part_hist_ptr[group_id * radix_range + j] = offset[j];
+                }
+            }
```

Comment:
part_hist_ptr is an gloabl mem ptr, therefore i would expect perf decrease from such parallelization because of increase of copy operations into global mem, but i will try your suggestion as part of optimization task.